src/main/java/io/snyk/agent/logic/ReportingWorker.java

```diff
@@ -132,6 +134,34 @@ void sendIfNecessary(Supplier<UseCounter.Drain> drainMaker) throws IOException {
         this.lastSuccessfulBeacon = now;
     }
 
+    private void setShutdownHook() {
+        try {
+            log.info("attempting to set up a shutdown hook that drains events one last time");
+            final Thread shutdownHook = new Thread(() -> safeSendIfNecessary(LandingZone.SEEN_SET::drain));
+            Runtime.getRuntime().addShutdownHook(shutdownHook);
+            log.info("shutdown hook set up successfully");
+        } catch (IllegalArgumentException e) {
+            // If the specified hook has already been registered,
+            // or if it can be determined that the hook is already running or has already been run.
+            // so either way we're set?
+        } catch (IllegalStateException e) {
+            // If the virtual machine is already in the process of shutting down
+            // this means we're okay with not setting up that hook
+        } catch (SecurityException e) {
+            // If a security manager is present and it denies RuntimePermission("shutdownHooks")
+            // not much to do about this one except log?
+            log.warn("failed to set a shutdownHook due to denied permission");
+        }
+    }
+
+    private void safeSendIfNecessary(Supplier<UseCounter.Drain> drainMaker) {
+        try {
+            sendIfNecessary(LandingZone.SEEN_SET::drain);
+        } catch (Throwable t) {
+            // intentionally disregard errors to allow fast shut down
```

Comment:
Catching here isn't super useful, we're in an independent thread, and any error which escapes the thread will just cause the thread to die anyway, perhaps with something else handling the exception (e.g. logging it properly).

If we do want to squash errors, we should probably only squash `(Exception ignored) {` here, catching `Throwable` is generally frowned upon.

Reply:
I'm not sure if I want this thread dying to somehow affect the actual app (for example if the app handles threads crashing).
changing from `throwable` to `exception` as you suggested, thanks 🙏 
what is the reasoning though, then, behind catching `throwable`s on the main loop at `run()`?

Reply:
The (rather weak) justification is:

1. The threads are workers, and don't want to die even if the job dies. They want to report the failure, but continue re-running the job. If we only saw `VmError`s (e.g. `OutOfMemoryError`), this probably wouldn't be a good enough justification, however...
2. Some of the VM APIs we're dealing with are more likely to throw `Error`s (i.e. the things that aren't `Exception`), due to being older, or weirder, than normal production Java code.

`FilterUpdate`'s case seems reasonable. 

That second point doesn't actually to the loop inside `ReportingWorker`, which I think I'd be in favour of changing to `catch (Exception`..?

`Tranformer` is re-throwing (i.e. not eating) the `Throwable`, which is more reasonable. We have a reasonable expectation that the code there isn't going to re-throw, even if the VM is in a bad state.